return err when (dis)connecting a non-exist network

allencloud opened the pull request:
Signed-off-by: allencloud <allen.sun@daocloud.io>

fixes https://github.com/docker/docker/issues/29754

**- What I did**
1. simplify function `ConnectToNetwork`, since we already got the network entity.
2. check network existence in `ConnectToNetwork `;
3. simplify function `DisconnectFromNetwork` into two kinds of cases, stopped container and running container
4. add test cases for the two above cases.

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

allencloud committed:
return err when (dis)connecting a non-exist network

Signed-off-by: allencloud <allen.sun@daocloud.io>

thaJeztah commented:
Although I can see this looking cleaner, I'm not sure this will work due to overlay / swarm scoped being extended automatically, so they may not be "present" yet when first checked; see [`daemon. findAndAttachNetwork ()`](https://github.com/docker/docker/blob/16500dfb1d1c0a3fc0d363ae782c22e8eda3b720/daemon/container_operations.go#L331-L411)

This change will error out before that code is reached, which probably is not what we want here

allencloud commented:
Oh, I think I have got your point.  @thaJeztah  However, actually I need a little bit more time to understand this network processing part. It is some kind of complex. If no enough background knowledge do I have, it would be hard.

allencloud commented:
@thaJeztah  At least, I think the change this PR brings to disconnect part is OK for us, right? :rofl:

allencloud commented on daemon/container_operations.go:
```diff
@@ -947,78 +947,56 @@ func errRemovalContainer(containerID string) error {
-				return err
-			}
-		} else {
-			container.NetworkSettings.Networks[idOrName] = &network.EndpointSettings{
```
I think here is the bug that when container is not running and we could not find the network, network name still has been written in container's network configs.

aboch replied:
Yes, for the connect to a non-running container not erroring out  issue.
Before https://github.com/docker/docker/pull/25962/commits/99a98ccc14a9427be47c8006e130750710db0a16, the network existence check was deferred to `daemon.updateNetworkConfig()`

aboch commented:
This work as expected on 1.12.5:
```
$ docker version | grep Version
 Version:      1.12.5
 Version:      1.12.5
$ docker create --name sleeping_cont busybox top
90e16067f75fd96b6b1ed5ca928be3587f803364da70691f7614c48f54b5a16a
$ docker network connect non-existent-nw sleeping_cont 
Error response from daemon: network non-existent-nw not found
$ docker run -d --name running_cont busybox top
aff51836085e95154cbbb729177633440bbd7e32db37c910f42b3e53e861d613
$ docker network connect non-existent-nw running_cont 
Error response from daemon: network non-existent-nw not found
$ 
```
Then it could be the breakage was introduced by the changes for supporting docker run on swarm attachable networks.

allencloud commented:
Updated this PR,  Not to influence findAndAttachNetwork(), just return error when container is not running and network does not exist.
PTAL @thaJeztah @aboch

allencloud commented:
@thaJeztah 

> At least, I think the change this PR brings to disconnect part is OK for us, right?

I do not think this opinion is right. The disconnecting part is also wrong. Because there is one case that we could disconnect a stopped from a remove network(since it is already removed, we could not find this network, while network details are still in the network config of this stopped container ).

This problem is a little bit complex, I will try to fix this, but I do not hope to see the increasing complexity to the code.

allencloud commented:
Hi, @thaJeztah @aboch 
I updated this PR to fix the related issue.  PTAL. Thanks.

aboch commented on daemon/container_operations.go:
```diff
@@ -957,68 +957,105 @@ func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName
-		if container.HostConfig.NetworkMode.IsHost() && containertypes.NetworkMode(n.Type()).IsHost() {
-			return runconfig.ErrConflictHostNetwork
+	if !container.Running {
+		if err := daemon.DisconnectStoppedContainerFromNetwork(container, networkName); err != nil {
```
```
{
    if !container.Running {
          return daemon.DisconnectStoppedContainerFromNetwork(container, networkName)
    }
    return daemon.DisconnectRunningContainerFromNetwork(container, networkName, force)
}
```

aboch commented on daemon/container_operations.go:
```diff
@@ -957,68 +957,105 @@ func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName
+	}
+
+	if !daemon.isNetworkHotPluggable() {
+		return fmt.Errorf(runtime.GOOS + " does not support connecting a running container to a network")
```
`connecting` -> `disconnecting`

allencloud replied:
Thanks. @aboch 
I updated the cases you mentioned, PTAL.

aboch commented on daemon/container_operations.go:
```diff
@@ -957,68 +957,101 @@ func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName
+			// if stopped container has no this non-found network details, return no such network error
 			return err
 		}
+		delete(container.NetworkSettings.Networks, networkName)
```
Have only one `delete(container.NetworkSettings.Networks, networkName)` after the if block.

aboch commented on daemon/container_operations.go:
```diff
@@ -957,68 +957,101 @@ func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName
+		if _, exist := container.NetworkSettings.Networks[networkName]; !exist {
+			return fmt.Errorf("container %s is not connected to the network %s", container.ID, networkName)
+		}
+		delete(container.NetworkSettings.Networks, networkName)
```
as above, remove this.

allencloud replied:
Thanks, removing is reasonable.

aboch commented:
@allencloud I agree with the correctness this fix is bringing, but I think this change will no longer allow connecting a stopped container to an attachable swarm network (if no tasks on that network are running yet on this node), because of the lazy swarm network creation.
We need to see if this can be a problem for some deployments.

allencloud commented:
@aboch 
Thanks for your important feedback. Both the original and current code this PR brings do not take **connect stopped container to lazy created network**

> We need to see if this can be a problem for some deployments.

👍

allencloud commented:
@aboch 
Any update with this one?
Should we move this on, or separate this into small ones to fix the bugs?

LK4D4 commented:
@aboch @sanimej @mavenugo can we review this?

aboch commented:
@allencloud Can you think of a way to not block the use case of connecting a container on a worker node to an attachable network which is not known to the node yet ?

allencloud commented:
Thanks for the your suggestion, and I will try that. Moving to WIP. @aboch

mlaventure commented:
ping @aboch @allencloud still a WIP?

allencloud commented:
Yeah, I have to say that it is still WIP.

allencloud committed:
not to block lazily created network on worker

Signed-off-by: allencloud <allen.sun@daocloud.io>

mlaventure commented:
ping @allencloud :angel:

thaJeztah commented:
```
05:41:32 # github.com/docker/docker/daemon
05:41:32 daemon/container_operations.go:1134: daemon.cluster.IsMnager undefined (type Cluster has no field or method IsMnager)
05:41:32 daemon/container_operations.go:1135: invalid type for composite literal: libnetwork.Network
05:41:32 daemon/container_operations.go:1143: invalid type for composite literal: libnetwork.Network
05:41:32 daemon/container_operations.go:1155: invalid type for composite literal: libnetwork.Network
```

allencloud commented:
@thaJeztah @aboch 
Sorry for my delay. Currently I have updated the PR.

> Can you think of a way to not block the use case of connecting a container on a worker node to an attachable network which is not known to the node yet ?

I added a function named `isNetworkLazyCreated` which shows whether a network is lazy created. So if a node try to find a network and the error is `libnetwork.ErrNoSuchNetwork`, then docker daemon will wait for 3 seconds to get the network if it is still gossiped.

PTAL, feel free to tell me if I could provide more details or update more. Thanks a lot.

cpuguy83 commented on daemon/container_operations.go:
```diff
@@ -1084,3 +1129,31 @@ func (daemon *Daemon) DeactivateContainerServiceBinding(containerName string) er
+			}
+
+			// network is not found, continue the loop
+			time.Sleep(1 * time.Second)
```
😨

allencloud replied:
I am afraid this is improper to hard code the duration. I think this is your concern. While this design needs more review from you maintainers.

mlaventure commented:
ping @aboch @cpuguy83 @allencloud  :innocent:

cpuguy83 commented:
I'm not sure on this one (literally).
While it's a nice idea to say the network is not found this can cause some issues as you can see from the code (gossipped networks) and can lead to race conditions.

cpuguy83 commented:
I'm going to close this one for now since there isn't any activity and no one has mentioned any sort of disagreement with my assessment from August.

Thanks! 🙇 😇
